_posts/2019-02-06-Age-Targeting-Part2.md

```diff
@@ -0,0 +1,255 @@
+grid_builder = grid_builder.addGrid(lr_model.elasticNetParam, params["elastic_net_params"])
+grid_builder = grid_builder.addGrid(lr_model.regParam, params["reg_params"])
+grid_builder = grid_builder.addGrid(lr_model.maxIter, params["max_iter"])
+
+evaluator = MulticlassClassificationEvaluator(labelCol=LABEL_COL, metricName="f1")
+cv = CrossValidator(
+    estimator=estimator,
+    estimatorParamMaps=grid_builder.build(),
+    evaluator=evaluator,
+    numFolds=5
+)
+cv_model = cv.fit(df)
+
+# Example to get the best parameters values:
+logging.info(
+    "regularisation parameter from best model: %f",
+    cv_model.bestModel.stages[-1]._java_obj.getRegParam() # work-around for Spark bug
+)
+best_model = cv_model.bestModel
+```
+
+It is worth mentioning that the introduction and the tuning of the regularization parameters turned out to be the most crucial one for both performance and stability of the model.
+
+**Trust the model**<br/>
+We want to know, if the model and the ground truth are reliable.
+Therefore, we have a look into the weights of the trained Logistic Regression Model.
+These weights can be interpreted as a feature importance and the top ones are shown in the picture below.
+Thus we can observe the most relevant aspects of our data which allow the model to discover the age behind a certain device.
+
+|age_group|weight|app description                                |
+|---------|------|---------------------------------------------|
+|13-17    |3.22  |discover the meaning of my name                |
+|13-17    |2.34  |math operation to solve homework exercise      |
+|13-17    |2.29  |famous birthdays                               |
+|18-24    |2.65  |game with quiz                                 |
+|18-24    |2.54  |test your competence                           |
+|18-24    |2.19  |test simulation                                |
+|25-34    |2.53  |chat, meet and date                            |
+|25-34    |2.24  |financial balance check                        |
+|25-34    |1.91  |calculator                                     |
+|35-54    |2.25  |videogames                                     |
+|35-54    |2.24  |photo collage                                  |
+|35-54    |1.73  |sport tracking (bike, jogging)                 |
+|55+      |2.14  |news and wether forecast                       |
+|55+      |2.08  |solitarie card games                           |
+|55+      |2.04  |solitarie card games                          |
+
+As you can see the best features of each class are the apps used.
+Due to the bag of words preprocess on this feature, every apps is a distinct input, so the weights in the table are directly related to the discrimination power derived by the usage of that app.
+Moreover, the list make extremely sense with what you would expect.
+This gives additional confidence on putting the model into production.
+Solve your math homework is something which tedius your life only when you are young.
```

Comment:
Solving your math homework is mostly a task you want to support with an app when you are young.